### PR TITLE
Add support for ip_range datatype (Backport of #3202 to 5.x)

### PR DESCRIPTION
**Pull request by codebrain**
This commit adds support for the ip_range datatype.
The existing IpRange type used for IpRangeAggregation renamed to
IpRangeAggregationRange to allow for the introduction of an IpRange type
whose name aligns with other range datatypes.

### CODE REVIEW

File: src/Nest/Aggregations/Bucket/IpRange/IpRangeAggregation.cs [hunk 1]
```diff
@@ -13,14 +13,14 @@ public interface IIpRangeAggregation : IBucketAggregation
 		[JsonProperty("field")]
 		Field Field { get; set; }
 
-		[JsonProperty(PropertyName = "ranges")]
-		IEnumerable<IIpRange> Ranges { get; set; }
+		[JsonProperty("ranges")]
+		IEnumerable<IIpRangeAggregationRange> Ranges { get; set; }
```
Comment (russcam) [inline src/Nest/Aggregations/Bucket/IpRange/IpRangeAggregation.cs]: Revert rename; breaks BWC

File: src/Nest/Aggregations/Bucket/IpRange/IpRangeAggregation.cs [hunk 2]
```diff
@@ -13,14 +13,14 @@ public interface IIpRangeAggregation : IBucketAggregation
...
 	}
 
 	public class IpRangeAggregation : BucketAggregationBase, IIpRangeAggregation
 	{
 		public Field Field { get; set; }
-		public IEnumerable<IIpRange> Ranges { get; set; }
+		public IEnumerable<IIpRangeAggregationRange> Ranges { get; set; }
```
Comment (russcam) [inline src/Nest/Aggregations/Bucket/IpRange/IpRangeAggregation.cs]: Revert rename; breaks BWC

File: src/Nest/Aggregations/Bucket/IpRange/IpRangeAggregation.cs [hunk 3]
```diff
@@ -36,13 +36,13 @@ public class IpRangeAggregationDescriptor<T> :
 	{
 		Field IIpRangeAggregation.Field { get; set; }
 
-		IEnumerable<IIpRange> IIpRangeAggregation.Ranges { get; set; }
+		IEnumerable<IIpRangeAggregationRange> IIpRangeAggregation.Ranges { get; set; }
```
Comment (russcam) [inline src/Nest/Aggregations/Bucket/IpRange/IpRangeAggregation.cs]: Revert rename; breaks BWC

File: src/Nest/Aggregations/Bucket/IpRange/IpRangeAggregation.cs [hunk 4]
```diff
@@ -36,13 +36,13 @@ public class IpRangeAggregationDescriptor<T> :
...
 
 		public IpRangeAggregationDescriptor<T> Field(Field field) => Assign(a => a.Field = field);
 
 		public IpRangeAggregationDescriptor<T> Field(Expression<Func<T, object>> field) => Assign(a => a.Field = field);
 
-		public IpRangeAggregationDescriptor<T> Ranges(params Func<IpRangeDescriptor, IIpRange>[] ranges) =>
-			Assign(a => a.Ranges = ranges?.Select(r => r(new IpRangeDescriptor())));
+		public IpRangeAggregationDescriptor<T> Ranges(params Func<IpRangeAggregationRangeDescriptor, IIpRangeAggregationRange>[] ranges) =>
```
Comment (russcam) [inline src/Nest/Aggregations/Bucket/IpRange/IpRangeAggregation.cs]: Revert rename; breaks BWC

File: src/Nest/Aggregations/Bucket/IpRange/IpRangeAggregation.cs [hunk 5]
```diff
@@ -36,13 +36,13 @@ public class IpRangeAggregationDescriptor<T> :
...
+			Assign(a => a.Ranges = ranges?.Select(r => r(new IpRangeAggregationRangeDescriptor())));
```
Comment (russcam) [inline src/Nest/Aggregations/Bucket/IpRange/IpRangeAggregation.cs]: Revert rename; breaks BWC

File: src/Nest/Mapping/Types/Core/Range/Ip/IpRangeAttribute.cs [hunk 6]
```diff
@@ -0,0 +1,10 @@
+﻿using System;
+using Elasticsearch.Net;
+
+namespace Nest
+{
+	public class IpRangeAttribute : RangePropertyAttributeBase, IIpRangeProperty
```
Comment (russcam) [inline src/Nest/Mapping/Types/Core/Range/Ip/IpRangeAttribute.cs]: `IpAddressRangeAttribute`, to align with `IpAddressRange`?

File: src/Nest/Mapping/Types/Core/Range/Ip/IpRangeProperty.cs [hunk 7]
```diff
@@ -0,0 +1,21 @@
+using System;
+using Elasticsearch.Net;
+using Newtonsoft.Json;
+
+namespace Nest
+{
+	[JsonObject(MemberSerialization.OptIn)]
+	public interface IIpRangeProperty : IRangeProperty { }
```
Comment (russcam) [inline src/Nest/Mapping/Types/Core/Range/Ip/IpRangeProperty.cs]: `IIpAddressRangeProperty`?

File: src/Nest/Mapping/Types/Core/Range/Ip/IpRangeProperty.cs [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+	public class IpRangeProperty : RangePropertyBase, IIpRangeProperty
```
Comment (russcam) [inline src/Nest/Mapping/Types/Core/Range/Ip/IpRangeProperty.cs]: `IpAddressRangeProperty `?

File: src/Nest/Mapping/Types/Core/Range/Ip/IpRangeProperty.cs [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+	{
+		public IpRangeProperty() : base(RangeType.IpRange) { }
+	}
+
+	public class IpRangePropertyDescriptor<T>
```
Comment (russcam) [inline src/Nest/Mapping/Types/Core/Range/Ip/IpRangeProperty.cs]: `IpAddressRangePropertyDescriptor`?

File: src/Nest/Mapping/Types/Properties.cs [hunk 10]
```diff
@@ -2,6 +2,8 @@
 using System.Collections.Generic;
 using System.Linq;
 using System.Linq.Expressions;
+using System.Reflection;
+using Elasticsearch.Net;
```
Comment (russcam) [inline src/Nest/Mapping/Types/Properties.cs]: needed?

File: src/Tests/Mapping/Types/Core/Range/IpRange/IpRangeAttributeTests.cs [hunk 11]
```diff
@@ -0,0 +1,26 @@
+﻿using Nest;
+using Tests.Framework;
+
+namespace Tests.Mapping.Types.Core.Range.IpRange
+{
+	public class IpRangeTest
```
Comment (russcam) [inline src/Tests/Mapping/Types/Core/Range/IpRange/IpRangeAttributeTests.cs]: `IpAddressRangeTest` for consistency?

File: src/Tests/Mapping/Types/Core/Range/IpRange/IpRangeAttributeTests.cs [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+	{
+		[IpAddressRange]
+		public Nest.IpAddressRange Range { get; set; }
+	}
+
+	[SkipVersion("<5.5.0", "ip range type is a new 5.5.0 feature")]
+	public class IpRangeAttributeTests : AttributeTestsBase<IpRangeTest>
```
Comment (russcam) [inline src/Tests/Mapping/Types/Core/Range/IpRange/IpRangeAttributeTests.cs]: `IpAddressRangeAttributeTests ` for consistency?

File: src/Tests/Mapping/Types/Core/Range/IpRange/IpRangePropertyTests.cs [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+using Nest;
+using Tests.Framework;
+using Tests.Framework.Integration;
+using Tests.Framework.ManagedElasticsearch.Clusters;
+using Tests.Framework.MockData;
+
+namespace Tests.Mapping.Types.Core.Range.IpRange
+{
+	[SkipVersion("<5.5.0", "ip range type is a new 5.5.0 feature")]
+	public class IpRangePropertyTests : PropertyTestsBase
```
Comment (russcam) [inline src/Tests/Mapping/Types/Core/Range/IpRange/IpRangePropertyTests.cs]: `IpAddressRangePropertyTests`  for consistency?

Comment (codebrain) [reply src/Tests/Mapping/Types/Core/Range/IpRange/IpRangePropertyTests.cs]: :-1:

File: src/Tests/Mapping/Types/Core/Range/IpRange/IpRangeAttributeTests.cs (hunk 12 above)
Comment (codebrain) [reply src/Tests/Mapping/Types/Core/Range/IpRange/IpRangeAttributeTests.cs]: :-1:

File: src/Tests/Mapping/Types/Core/Range/IpRange/IpRangeAttributeTests.cs (hunk 11 above)
Comment (codebrain) [reply src/Tests/Mapping/Types/Core/Range/IpRange/IpRangeAttributeTests.cs]: :-1:

File: src/Nest/Mapping/Types/Core/Range/Ip/IpRangeProperty.cs (hunk 9 above)
Comment (codebrain) [reply src/Nest/Mapping/Types/Core/Range/Ip/IpRangeProperty.cs]: :-1:

File: src/Nest/Mapping/Types/Core/Range/Ip/IpRangeProperty.cs (hunk 7 above)
Comment (codebrain) [reply src/Nest/Mapping/Types/Core/Range/Ip/IpRangeProperty.cs]: :-1:

File: src/Nest/Mapping/Types/Core/Range/Ip/IpRangeProperty.cs (hunk 8 above)
Comment (codebrain) [reply src/Nest/Mapping/Types/Core/Range/Ip/IpRangeProperty.cs]: :-1:

File: src/Nest/Mapping/Types/Core/Range/Ip/IpRangeAttribute.cs (hunk 6 above)
Comment (codebrain) [reply src/Nest/Mapping/Types/Core/Range/Ip/IpRangeAttribute.cs]: :-1:

**Pull request merged**